src/aws-fsx/ontap-file-system.ts

```diff
@@ -0,0 +1,520 @@
+   * First-generation file systems are powered by 1 HA pair.
+   * Second-generation multi-AZ file systems are powered by 1 HA pair.
+   * Second generation single-AZ file systems are powered by up to 12 HA pairs.
+   *
+   * The value of this property affects the values of StorageCapacity, Iops, and ThroughputCapacity.
+   *
+   * Block storage protocol support (iSCSI and NVMe over TCP) is disabled on file systems with more than 6 HA pairs.
+   *
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/administering-file-systems.html#HA-pairs
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/supported-fsx-clients.html#using-block-storage
+   *
+   * @default - 1 HA pair
+   */
+  readonly haPairs?: number;
+
+  /**
+   * The subnet in which you want the preferred file server to be located.
+   *
+   * This value is required when DeploymentType is set to MULTI_AZ_1 or MULTI_AZ_2.
+   *
+   * @default - no default value (This value is not used for single-AZ file systems, but it is required for multi-AZ file systems)
+   */
+  readonly preferredSubnet?: aws_ec2.ISubnet;
+
+  /**
+   * The route tables in which Amazon FSx creates the rules for routing traffic to the correct file server.
+   * You should specify all virtual private cloud (VPC) route tables associated with the subnets in which your clients are located.
+   *
+   * Amazon FSx manages VPC route tables for Multi-AZ file systems using tag-based authentication.
+   * These route tables are tagged with Key: AmazonFSx; Value: ManagedByAmazonFSx.
+   *
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/unable-to-access.html#vpc-route-tables-not-tagged
+   *
+   * @default - Amazon FSx selects your VPC's default route table.
+   */
+  readonly routeTables?: aws_ec2.IRouteTable[];
+
+  /**
+   * The throughput capacity for the file system that you're creating in megabytes per second (MBps).
+   *
+   * You can define either the `throughputCapacityPerHaPair` or the `throughputCapacity` when creating a file system, but not both.
+   *
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/managing-throughput-capacity.html
+   *
+   * @default - recommended throughput capacity based on the storage capacity
+   */
+  readonly throughputCapacity?: number;
+
+  /**
+   * The throughput capacity per HA pair, rather than the total throughput for the file system.
+   *
+   * You can define either the `throughputCapacityPerHaPair` or the `throughputCapacity` when creating a file system, but not both.
+   *
+   * For SINGLE_AZ_1 and MULTI_AZ_1 file systems, valid values are 128, 256, 512, 1024, 2048, or 4096 MBps.
+   * For SINGLE_AZ_2, valid values are 1536, 3072, or 6144 MBps.
+   * For MULTI_AZ_2, valid values are 384, 768, 1536, 3072, or 6144 MBps.
+   *
+   * @default - recommended throughput capacity based on the storage capacity
+   */
+  readonly throughputCapacityPerHaPair?: number;
```

Comment:
Could it be expressed using the Enum-Like class? This is a very complex and exclusive option, so I think it would be a match. (This is just an idea, I could be wrong though.)

If this is possible, it would be even cooler to implement if the validation rules could be confined within the class. (I don't think it to work that easily, but...)

Reply:
I've implemented `ThroughputCapacityPerHaPair` class.

Reply:
I have also implemented it. I think this is safer and validation can be closed within the class. What do you think?

```ts
export abstract class ThroughputCapacityPerHaPair {
  protected static calculateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number): number {
    if (haPairs <= 0 || !Number.isInteger(haPairs)) {
      throw new Error(`'haPairs' must be a positive integer, got ${haPairs}`);
    }
    return throughputCapacity / haPairs;
  }

  public abstract readonly allowedCapacity: number[];
  public abstract readonly deploymentType: OntapDeploymentType;

  protected constructor(public readonly capacity: number) {}
  protected validateCapacity() {
    if (!this.allowedCapacity.includes(this.capacity)) {
      throw new Error(
        `'throughputCapacity / haPairs' must be one of the following values for ${
          this.deploymentType
        }: ${this.allowedCapacity.join(', ')}. got: ${this.capacity} MB/s/HA pair`,
      );
    }
  }
}

export class SingleAz1ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_128 = new SingleAz1ThroughputCapacityPerHaPair(128);
  public static readonly MB_PER_SEC_256 = new SingleAz1ThroughputCapacityPerHaPair(256);
  public static readonly MB_PER_SEC_512 = new SingleAz1ThroughputCapacityPerHaPair(512);
  public static readonly MB_PER_SEC_1024 = new SingleAz1ThroughputCapacityPerHaPair(1024);
  public static readonly MB_PER_SEC_2048 = new SingleAz1ThroughputCapacityPerHaPair(2048);
  public static readonly MB_PER_SEC_4096 = new SingleAz1ThroughputCapacityPerHaPair(4096);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new SingleAz1ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();
  }

  public readonly allowedCapacity = [128, 256, 512, 1024, 2048, 4096];
  public readonly deploymentType = OntapDeploymentType.SINGLE_AZ_1;
}

export class MultiAz1ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_128 = new MultiAz1ThroughputCapacityPerHaPair(128);
  public static readonly MB_PER_SEC_256 = new MultiAz1ThroughputCapacityPerHaPair(256);
  public static readonly MB_PER_SEC_512 = new MultiAz1ThroughputCapacityPerHaPair(512);
  public static readonly MB_PER_SEC_1024 = new MultiAz1ThroughputCapacityPerHaPair(1024);
  public static readonly MB_PER_SEC_2048 = new MultiAz1ThroughputCapacityPerHaPair(2048);
  public static readonly MB_PER_SEC_4096 = new MultiAz1ThroughputCapacityPerHaPair(4096);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new MultiAz1ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();

  }

  public readonly allowedCapacity = [128, 256, 512, 1024, 2048, 4096];
  public readonly deploymentType = OntapDeploymentType.MULTI_AZ_1;
}

export class SingleAz2ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_1536 = new SingleAz2ThroughputCapacityPerHaPair(1536);
  public static readonly MB_PER_SEC_3072 = new SingleAz2ThroughputCapacityPerHaPair(3072);
  public static readonly MB_PER_SEC_6144 = new SingleAz2ThroughputCapacityPerHaPair(6144);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new SingleAz2ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();

  }

  public readonly allowedCapacity = [1536, 3072, 6144];
  public readonly deploymentType = OntapDeploymentType.SINGLE_AZ_2;
}

export class MultiAz2ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_384 = new MultiAz2ThroughputCapacityPerHaPair(384);
  public static readonly MB_PER_SEC_768 = new MultiAz2ThroughputCapacityPerHaPair(768);
  public static readonly MB_PER_SEC_1536 = new MultiAz2ThroughputCapacityPerHaPair(1536);
  public static readonly MB_PER_SEC_3072 = new MultiAz2ThroughputCapacityPerHaPair(3072);
  public static readonly MB_PER_SEC_6144 = new MultiAz2ThroughputCapacityPerHaPair(6144);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new MultiAz2ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();

  }

  public readonly allowedCapacity = [384, 768, 1536, 3072, 6144];
  public readonly deploymentType = OntapDeploymentType.MULTI_AZ_2;
}

// ...
// ...

private validateThroughputCapacity(
  throughputCapacity?: number,
  throughputCapacityPerHaPair?: ThroughputCapacityPerHaPair,
  haPair: number = 1,
): void {
  // ...
  // ...
  if (throughputCapacityPerHaPair.deploymentType !== this.deploymentType) {
    throw new Error('Error msg')
  }
  // ...
  // ...
  switch (this.deploymentType) {
    case OntapDeploymentType.SINGLE_AZ_1:
      SingleAz1ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    case OntapDeploymentType.MULTI_AZ_1:
      MultiAz1ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    case OntapDeploymentType.SINGLE_AZ_2:
      SingleAz2ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    case OntapDeploymentType.MULTI_AZ_2:
      MultiAz2ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
  }
}
```

Reply:
Thank a lot!! It is really nice to implement `{DeploymentType}ThroughputCapacityPerHaPair` for each. I'll update my code like this later.

One more point, I’m considering removing the `throughputCapacity` argument. The reason is that while proper validation is applied for `throughputCapacityPerHaPair`, `throughputCapacity` allows any number, which seems like a missed opportunity.

We could define a similar class for `throughputCapacity`, but I believe using just `throughputCapacityPerHaPair` would be sufficient. In fact, having both options might cause confusion for users. What are your thoughts on this?

Reply:
> One more point, I’m considering removing the throughputCapacity argument.

Where are you talking about? (Is it `validateThroughputCapacityPerHaPair` method?)

Is it that the validation for `throughputCapacity` and `haPairs` is not necessary? If the value of `throughputCapacity / haPairs` is not included in allowed capacity patterns, doesn't CFn fail?

Reply:
Sorry for the confusion. What I was referring to was not the validate function, but rather the idea of removing `throughputCapacity` from the `OntapConfiguration`.

It was quite unclear... my apologies for that.

What I wanted tosay is that `throughputCapacity` and `throughputCapacityPerHaPair` have very similar roles, and with the recent improvements, `throughputCapacityPerHaPair` has become much more user-friendly. Therefore, I believe it would be simpler to remove `throughputCapacity` altogether.

Reply:
I see. If there is no use case for `throughputCapacity`, I think it's a good idea too!

Reply:
L2 does not necessarily have to address all properties and simplicity is a good thing.

Reply:
I've updated my code including README, integ test, jsdoc like below.

```ts
export abstract class ThroughputCapacityPerHaPair {
  /**
   * The deployment type of the throughput capacity.
   */
  public abstract readonly deploymentType: OntapDeploymentType;
  protected abstract readonly allowedCapacity: number[];
  protected constructor(public readonly capacity: number) {}

  /**
   * Check if the throughput capacity is valid for the given deployment type.
   */
  public validateCapacity(): void {
    if (!this.allowedCapacity.includes(this.capacity)) {
      throw new Error(
        `'throughputCapacity / haPairs' must be one of the following values for ${
          this.deploymentType
        }: ${this.allowedCapacity.join(', ')}. got: ${this.capacity} MB/s/HA pair`,
      );
    }
  }
}

export class SingleAz2ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  /**
   * The throughput capacity of 1536 MBps per HA pair.
   */
  public static readonly MB_PER_SEC_1536 = new SingleAz2ThroughputCapacityPerHaPair(1536);

  /**
   * The throughput capacity of 3072 MBps per HA pair.
   */
  public static readonly MB_PER_SEC_3072 = new SingleAz2ThroughputCapacityPerHaPair(3072);

  /**
   * The throughput capacity of 6144 MBps per HA pair.
   */
  public static readonly MB_PER_SEC_6144 = new SingleAz2ThroughputCapacityPerHaPair(6144);

  /**
   * The deployment type of the throughput capacity.
   */
  public readonly deploymentType = OntapDeploymentType.SINGLE_AZ_2;
  protected readonly allowedCapacity = [1536, 3072, 6144];
}
```

And just now, I think `ThroughputCapacityPerHaPair.validateCapacity()` may be unnecessary because all of the `throughputCapacity` is configured by `DeploymentTypeThroughputCapacityPerHaPair`. The validation of mismatch of deployment type between `deploymenType` prop and `throughputCapacity` prop is already executed at [this line](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R627). Therefore, I think it is better to remove `ThroughputCapacityPerHaPair.validateCapacity()` and [this validation](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R632).

Reply:
> And just now, I think ThroughputCapacityPerHaPair.validateCapacity() may be unnecessary because all of the throughputCapacity is configured by DeploymentTypeThroughputCapacityPerHaPair. The validation of mismatch of deployment type between deploymenType prop and throughputCapacity prop is already executed at [this line](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R627). Therefore, I think it is better to remove ThroughputCapacityPerHaPair.validateCapacity() and [this validation](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R632).

I think so!